### ISSUE TITLE
Expected loss during training

### ISSUE THREAD

**Issue by modernAlcibiades** [rwth-i6/returnn-experiments#10]
I am trying to define a new loss based on WER as per this paper - https://arxiv.org/pdf/1712.01818.pdf. Is there a way to run beam search at train time, calculate WER using edit_distance and incorporate that as expected_loss into the training ce loss?

**Comment by albertz** [comment]
Yes, I have done this already. I think you should find a Max expected Bleu
training config for machine translation here on this repo.


Am Do., 8. Nov. 2018, 13:54 hat modernAlcibiades <notifications@github.com>
geschrieben:

**Comment by modernAlcibiades** [comment]
Hi
Are you referring to [this](https://github.com/rwth-i6/returnn-experiments/blob/9ee943d8c663419208102fa45a9ad5734583d55d/2018-attention/wmt2017deen/mbleu.enc4l.pretrain2.adam.lr1e_3.mseqs100.bs4000.ls01.tembi0.invfert.oeps1e_8.gradnoise0.seqsort1000.config)

My understanding is that ce loss in output_prob is not calculated when search_train_layers is set. Is that correct?
I want to be able to train using both ce loss, as well as bleu score for the hypothesis. Is there a way to merge bleu score with ce loss, ie run without setting 'loss_only_on_non_search' parameter in output_prob?

**Comment by albertz** [comment]
Yes exactly, this is one example config.

When search_train_layers is set, it will build a second network for the
specified layers where it does search. In the original network, search is
disabled and you will get the CE loss. In the second search network, search
is enabled. By setting the loss scales, you can weighten both losses as you
want.
Best see the code for this specific option to see how it works.

Best,
Albert



Am Do., 8. Nov. 2018, 17:46 hat modernAlcibiades <notifications@github.com>
geschrieben:
